apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsLinkControl.tsx

```diff
@@ -0,0 +1,198 @@
+import type { UseRadioProps } from "@chakra-ui/react"
+import type { ControlProps, RankedTester } from "@jsonforms/core"
+import type { PropsWithChildren } from "react"
+import { useState } from "react"
+import {
+  Box,
+  FormControl,
+  HStack,
+  Icon,
+  Text,
+  useRadio,
+  useRadioGroup,
+} from "@chakra-ui/react"
+import { and, isStringControl, rankWith, schemaMatches } from "@jsonforms/core"
+import { withJsonFormsControlProps } from "@jsonforms/react"
+import { FormLabel, Input } from "@opengovsg/design-system-react"
+import { BiEnvelopeOpen, BiFile, BiFileBlank, BiLink } from "react-icons/bi"
+
+import { JSON_FORMS_RANKING } from "~/constants/formBuilder"
+
+const LINK_TYPES = {
+  page: {
+    icon: BiFileBlank,
+    label: "Page",
+  },
+  external: {
+    icon: BiLink,
+    label: "External link",
+  },
+  file: {
+    icon: BiFile,
+    label: "File",
+  },
+  email: {
+    icon: BiEnvelopeOpen,
+    label: "Email",
+  },
+} as const
+
+export const jsonFormsLinkControlTester: RankedTester = rankWith(
+  JSON_FORMS_RANKING.LinkControl,
+  and(
+    isStringControl,
+    schemaMatches((schema) => schema.format === "link"),
+  ),
+)
+
+const RadioCard = ({ children, ...rest }: PropsWithChildren<UseRadioProps>) => {
+  const { getInputProps, getRadioProps } = useRadio(rest)
+
+  return (
+    <Box
+      as="label"
+      _first={{
+        "> div": {
+          borderLeftRadius: "base",
+        },
+      }}
+      _last={{
+        "> div": {
+          borderRightRadius: "base",
+        },
+      }}
+    >
+      <input {...getInputProps()} />
+      <Box
+        {...getRadioProps()}
+        cursor="pointer"
+        border="1px solid"
+        borderColor="base.divider.strong"
+        bgColor="utility.ui"
+        px="1rem"
+        py="0.5rem"
+        mx={0}
+        _checked={{
+          bgColor: "interaction.muted.main.active",
+          color: "interaction.main.default",
+          borderColor: "interaction.main.default",
+        }}
+        textTransform="none"
+        fontWeight={500}
+        lineHeight="1.25rem"
+      >
+        {children}
+      </Box>
+    </Box>
+  )
+}
+
+interface RadioContentProps {
+  selectedLinkType: string
+  data: string
+  handleChange: (value: string) => void
+}
+
+const RadioContent = ({
+  selectedLinkType,
+  data,
+  handleChange,
+}: RadioContentProps) => {
```

Comment:
assert return type for typechecking

```suggestion
}: RadioContentProps): JSX.Element => {
```

Reply:
Fixed in 5e33dc3f352940442a3b6f798764b8dd3fef94d1.